File: native-ios/RideTrackerNative/ExternalTelemetryGateway.swift
import Foundation

struct ExternalTelemetryChannel: Codable, Hashable {
    let metric: String
    let channelID: String
    let value: Double
    let quality: Double
}

struct ExternalTelemetryPacket: Codable, Hashable {
    let deviceID: String
    let timestamp: TimeInterval
    let channels: [ExternalTelemetryChannel]
}

@MainActor
final class ExternalTelemetryGateway {
    private let router: TelemetrySourceRouter

    init(router: TelemetrySourceRouter) {
        self.router = router
    }

    func ingest(_ packet: ExternalTelemetryPacket) {
        let timestamp = packet.timestamp > 0 ? packet.timestamp : ProcessInfo.processInfo.systemUptime
        for channel in packet.channels where channel.value.isFinite {
            router.ingest(
                metric: channel.metric,
                sourceID: "\(packet.deviceID)/\(channel.channelID)",
                value: channel.value,
                quality: max(0, min(1, channel.quality)),
                timestamp: timestamp
            )
        }
    }

    func ingestAccessorySample(_ sample: BLEAccessoryManager.AccessorySample, deviceID: String = "external-gnss") {
        var channels: [ExternalTelemetryChannel] = []
        let quality = max(0, min(1, sample.quality ?? sample.horizontalAccuracy.map { 1 - min($0, 100) / 100 } ?? 1))
        if let speed = sample.speedMS, speed.isFinite {
            channels.append(.init(metric: "speedKmh", channelID: "speed", value: max(0, speed * 3.6), quality: quality))
        }
        if let x = sample.accelerationX, x.isFinite { channels.append(.init(metric: "accelerationX", channelID: "accelerationX", value: x, quality: quality)) }
        if let y = sample.accelerationY, y.isFinite { channels.append(.init(metric: "accelerationY", channelID: "accelerationY", value: y, quality: quality)) }
        if let z = sample.accelerationZ, z.isFinite { channels.append(.init(metric: "accelerationZ", channelID: "accelerationZ", value: z, quality: quality)) }
        ingest(.init(deviceID: deviceID, timestamp: ProcessInfo.processInfo.systemUptime, channels: channels))
    }
}

File: native-android/app/src/main/java/de/ridetracker/sensors/ExternalTelemetryGateway.kt
package de.ridetracker.sensors

import android.os.SystemClock

data class ExternalTelemetryChannel(
    val metric:String,
    val channelId:String,
    val value:Double,
    val quality:Double = 1.0,
)

data class ExternalTelemetryPacket(
    val deviceId:String,
    val timestampMs:Long = SystemClock.elapsedRealtime(),
    val channels:List<ExternalTelemetryChannel>,
)

class ExternalTelemetryGateway(private val router:TelemetrySourceRouter) {
    fun ingest(packet:ExternalTelemetryPacket) {
        val timestamp=if(packet.timestampMs>0)packet.timestampMs else SystemClock.elapsedRealtime()
        packet.channels.filter{it.value.isFinite()}.forEach{channel->
            router.ingest(
                metric=channel.metric,
                sourceId="${packet.deviceId}/${channel.channelId}",
                value=channel.value,
                quality=channel.quality.coerceIn(0.0,1.0),
                timestampMs=timestamp,
            )
        }
    }

    fun ingestGnssSpeed(speedMs:Double,quality:Double=1.0,deviceId:String="external-gnss") {
        if(!speedMs.isFinite())return
        ingest(ExternalTelemetryPacket(deviceId=deviceId,channels=listOf(ExternalTelemetryChannel("speedKmh","speed",speedMs.coerceAtLeast(0.0)*3.6,quality))))
    }

    fun ingestHeartRate(bpm:Int,quality:Double=1.0,deviceId:String="external-heart") {
        if(bpm<=0)return
        ingest(ExternalTelemetryPacket(deviceId=deviceId,channels=listOf(ExternalTelemetryChannel("heartRateBpm","heartRate",bpm.toDouble(),quality))))
    }

    fun ingestImu(x:Double,y:Double,z:Double,quality:Double=1.0,deviceId:String="external-imu") {
        ingest(ExternalTelemetryPacket(deviceId=deviceId,channels=listOf(
            ExternalTelemetryChannel("accelerationX","accelerationX",x,quality),
            ExternalTelemetryChannel("accelerationY","accelerationY",y,quality),
            ExternalTelemetryChannel("accelerationZ","accelerationZ",z,quality),
        )))
    }
}
